File: reverseknodes.java
// Time complexity - O(n)
// Space complexity - O(1)
// Works on leetcode

class Solution {
    private ListNode reverse(ListNode begin, ListNode end){
        ListNode prev = begin;
        ListNode current = begin.next;
        ListNode next = current.next;
        ListNode head = begin.next;
        
        while(next != end){
            current.next = prev;
            prev = current;
            current = next;
            next = next.next;
        }
        current.next = prev;
        begin.next = current;
        head.next = end;
        return head;
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
      if(head == null){
          return head;
      } 
      ListNode dummy = new ListNode(-1);
      dummy.next = head;
      ListNode begin = dummy;
      ListNode first = null;
      ListNode current = head;
      int i = 0; 
      while(current != null){
          i = i + 1;
          if(i % k == 0){
              first = reverse(begin, current.next);
              begin = first;
              current = first.next;
          }
          else{
              current = current.next;
          }
      }    
      return dummy.next;
    }
}

File: reverseKNodes.py
#Time: O(n)
#Space: O(1)
#Program ran on leetcode successfully

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        
        if not head or not head.next:
            return head

        tail = head
        for i in range(k):
            if not tail:
                return head

            tail = tail.next

        tail = self.reverseKGroup(tail, k)

        for i in range(k):
            next = head.next
            head.next = tail
            tail = head
            head = next

        return tail